rfcs/0042-config-option.md

```diff
@@ -0,0 +1,301 @@
+- Responsibility for backwards compatibility: By adding such options, we obligate ourselves to handle backwards incompatibilites on our side. We will have to support these options for a long time and can't remove them without at least one person being annoyed about it.
+
+## `configFile` or `extraConfig` option
+
+An option for specifying the contents of the configuration file directly with `configFile` or `extraConfig`. Problems:
+
+- Not very modular at all
+  - In case of json or a `configFile` option, you can only assign the option once, merging is impossible
+  - In case of ini (or similar), assigning a single option multiple times to make use of list concatenation, ordering or priorities is impossible, so in general you can't e.g. override a default value set somewhere else.
+- No syntax checking: Users will have to know the syntax of the configuration language and encode their values properly, any syntax error will only be realized when the program errors at runtime.
+
+## Occurences of problems
+
+- Because prometheus uses options to encode every possible setting, [#56017](https://github.com/NixOS/nixpkgs/pull/56017) is needed to allow users to set a part of the configuration that wasn't encoded yet.
+- Because strongswan-ctl uses options to encode its full configuration, changes like [#49197](https://github.com/NixOS/nixpkgs/pull/49197) are needed to update our options with upstream changes.
+- Pull requests like [#57036](https://github.com/NixOS/nixpkgs/pull/57036) or [#38324](https://github.com/NixOS/nixpkgs/pull/38324) are needed because users wish to have more configuration options than the ones provided.
+- [#58239](https://github.com/NixOS/nixpkgs/pull/58239), [#58181](https://github.com/NixOS/nixpkgs/pull/58181)
+
+## Previous discussions
+
+- https://github.com/NixOS/nixpkgs/pull/44923#issuecomment-412393196
+- https://github.com/NixOS/nixpkgs/pull/55957#issuecomment-464561483 -> https://github.com/NixOS/nixpkgs/pull/57716
+
+## Previous implementations
+
+This idea has been implemented already in some places:
+- [#45470](https://github.com/NixOS/nixpkgs/pull/45470)
+- [#52096](https://github.com/NixOS/nixpkgs/pull/52096)
+- [My Murmur module](https://github.com/Infinisil/system/blob/45c3ea36651a2f4328c8a7474148f1c5ecb18e0a/config/new-modules/murmur.nix)
+
+# Detailed design
+[design]: #detailed-design
+
+For specifying configuration files to programs in NixOS options, there should be a main single option called `config` which represents the configuration of the program as a Nix value, which can then be converted to the configuration file format the program expects. This may look as follows:
+
+```nix
+{ config, lib, ... }: with lib;
+let
+
+  cfg = config.services.foo;
+
+  configText = configGen.json cfg.config;
+
+in {
+
+  options.services.foo = {
+    enable = mkEnableOption "foo service";
+
+    config = mkOption {
```

Comment:
The name `config` is quite overloaded. I don't know whether the different uses of the term could cause confusion.

In HM we sometimes use the term `settings` for this use, for example, in the [`programs.noti`](https://github.com/rycee/home-manager/blob/f8b03f5750679d6686eefa60d5de8cfd932b8997/modules/programs/noti.nix#L17-L43) module. Unfortunately, we are not consistent and in some modules we also use the term `config` for this option 😖

Reply:
That's a good point, I haven't even considered other names. `config` is especially bad because it conflicts with the `config` in NixOS modules (`{ options = ...; config = ...; }`). I've also been using the term "settings" in this RFC for the programs settings already.

I've been trying to get submodules to support `imports = ` and `options = ` attributes, and options themselves being named `config` leads to ambiguity (If `foo` is an option of a submodule type which has `config = mkOption`, what should `foo.config =` mean? `foo.config = <value of the config option>` or `foo.config = <config section of the submodule>`).

I'm 80% for using "settings" instead of "config", only thing against it is that people are used to "config" already.

Reply:
I have added a TODO in https://github.com/NixOS/rfcs/pull/42/commits/e9082a92e0f97613cbdc9cb40ed0f6eb5a7ccd5e to discuss using `settings` instead.